models/solution/route.rb

```diff
@@ -55,6 +55,46 @@ def insert_stop(vrp, stop, index, idle_time = 0)
         shift_route_times(idle_time + stop.activity.duration, index)
       end
 
+      def remove_stop(vrp, index)
+        removed_stop = stops.delete_at(index)
+        previous_stop = stops[0..index - 1].reverse.find{ |n_stop| n_stop.activity.point } if index != 0
+        next_stop = stops[index..-1].find{ |n_stop| n_stop.activity.point } if index != stops.size
+        passed_stop = false
+        non_located_stops = stops[0..index - 1].reverse.select{ |stop|
+          passed_stop = true if stop == removed_stop
+          next if passed_stop
+
+          stop.activity.point.nil?
+        }
+        intermediate_duration = non_located_stops.sum{ |stop| stop.info.waiting_time + stop.activity.duration } +
+                                removed_stop.info.waiting_time
+
+        if previous_stop.type != :depot
+          raise 'Current implementation of remove stop should be only used in the context of removing the first service stop of a route'
+        end
+
+        next_start = next_stop&.info&.begin_time || 0
+        previous_matrix_index = previous_stop&.activity&.point&.matrix_index
+        next_matrix_index = next_stop&.activity&.point&.matrix_index
+        previous_end = previous_stop.info.end_time
+        route_matrix = vrp.matrices.find{ |matrix| matrix.id == vehicle.matrix_id }
+        if route_matrix.time
+          new_travel_time =
+            previous_matrix_index && next_matrix_index &&
+            route_matrix.time[previous_matrix_index][next_matrix_index] || 0
+          shift_route_times_backward(previous_end + intermediate_duration - next_start - new_travel_time, index)
```

Comment:
But (like you said) if there is force-start or if there are tw.end and/or tw.begin it might be preferable to push stuff to the left or right or both... how can we decide? 

here we assume, either we should add check at the beginning of this function so that the "simplificaiton"/"cleaning" is only done under strict (simple) conditions that we can handle and then we can improve this function (and relax the check) along the road. How does that sound?

We need to decide which simple case to handle first and do the realise with that case next week. Do we know the clients use case? Lets talk this Monday morning..

Reply:
In the context of the customer, we have both force_start and a timewindow on a service which stands for a located rest